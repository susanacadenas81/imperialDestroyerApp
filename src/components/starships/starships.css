.list-wrapper {
    display: flex;
    width: 90vw;
    flex-wrap: wrap;
}

.card{
    width: 25em;
    height: 20em;
    display: flex;
    flex-direction: column;
    border: 1px solid grey;
    margin: .5em;
    border-radius: 3px;
    background-color: lightgray;
}

.starship-image-wrapper {
    display: flex;
    justify-content: center;
    border: 1.2px solid rgb(184, 184, 184);
    width: 8em;
    float: left;
    border-radius: 2px;
    margin-top: .2em;
    margin-left: .2em;
    margin-right: .6em;
    background-color: black;
}

.starship-image {
    height: 8em;
    width: 7em;
    background-color: black;
}

.screen-wrapper{
    display: flex;
    max-width: 97vw;
    flex-wrap: wrap;
    margin-left: .5em;
    justify-content: center;
}

.title-icon {    
    width: .85em !important;
    height: .85em !important;
    margin-right: .2em;
    padding-right: .2em;
    padding-left: .2em;
    color: #333; 
}

.data-card {
    font-size: 1.2em;
    
}

.starships-section-title {
    display: flex;
    text-align: center;
    font-size: 2.1em;
    color: rgb(0, 0, 0);
}

.header-card {
    text-align: center;
    font-size: 1.2em;
    border-bottom: 1px solid grey;
    background-color: rgba(0, 0, 0, 0.89);
}

.title-card {
    font-family: 'Starjedi';
    color: rgba(255, 255, 0, 0.945);
    padding: 1em;
}

.data-card p {
    margin: .1em;
    margin-left: 1em;
    color: rgb(0, 0, 0);
}

.header-screen {
    padding: .3em;
    margin-left: .4em;
    width: 100%;
    max-width: 90.2vw;
    font-size: 1em;
    font-family: 'Starjedi';
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.dropdown {
    display: flex;
    align-items: center;
    margin-right: 5em;
}

.sort-icon {
    margin-right: .5em;
}

hr {
    margin-left: 1.5em;
    size: 1.5em;
}

@media (max-width: 850px) {
    .card {
        height: inherit !important;
        width: inherit !important;
        max-width: 25em;
    }
    hr {
        display: none;
    }
    .screen-wrapper {
        max-width: inherit;
        width: 120vw;
        justify-content: inherit;

    }
    .list-wrapper {
        width: 100%;
        justify-content: center;
        margin-left: 2em;
    }
}