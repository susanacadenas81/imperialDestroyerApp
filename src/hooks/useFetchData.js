import { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { getAllData } from "../helpers/getData";

export const useFetchData = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });    
    //const navigate = useNavigate();
    
    useEffect( () => {
        const callGetAllData = (category) => {
            const storageData = JSON.parse(localStorage.getItem(category));
            if (storageData && storageData.length) {
                setState({
                    count: storageData.length,
                    data: storageData,
                    loading: false
                }) 
            } else {
                getAllData({category}).then( data => {                
                    setState({
                        count: data.count,
                        data: data.results,
                        loading: false
                    }) 
                    }, err => { 
                        //navigate('/errorPage')
                    });
            }
        }
        
        if (category === 'starships') {
            callGetAllData('starships');
        } else {
            callGetAllData('planets')
        }

    }, [category])
    return state;

};