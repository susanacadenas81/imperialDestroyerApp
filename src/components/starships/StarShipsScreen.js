import Dropdown from 'react-dropdown';
import 'react-dropdown/style.css';
import { useState } from "react";
import { FaRocket, FaSortAmountDownAlt } from 'react-icons/fa'


import { StarshipList } from './StarshipList'
import './starships.css'

export const StarShipsScreen = () => {
  const [sortBy, setSortby] = useState('')
  const onSelect = (e) => {
    setSortby(() => e.value);
  }
  const sortOptions = [{value: 'cargo_capacity', label:'Capacity'}, {value:'crew', label:'Crew'}];
  return (
    <div className="screen-wrapper">
      <div className="header-screen">
        <h1 className="starships-section-title"><i><FaRocket className='title-icon' /></i>Starships</h1>
        <hr color='grey' width='65%' />
        <div className="dropdown">
          <i className="sort-icon"><FaSortAmountDownAlt/></i>
          <Dropdown options={sortOptions} onChange={onSelect} placeholder="Sort" />
        </div>
      </div>
        <StarshipList sortBy={sortBy} />
    </div>
  )
}
