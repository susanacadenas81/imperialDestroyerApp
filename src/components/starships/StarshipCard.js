import { formatBigNumbers } from '../../helpers/numbers';
//import '../styles/components.css';
import './starships.css';
import PropTypes from 'prop-types';

export const StarshipCard = ({ shipName, shipCapacity, MGLT, consumables, costInCredits, crew, hyperDriverRating, manufacturer }) => {
  const name = shipName.toLowerCase().replace(' ', '');
  const imagePath = `/assets/starships/${name}.png`
  const change2DefaultImage = (event) => {
    event.target.src = 'assets/default.webp';
  }
  return (
    <div className='card'>
        <div className='data-wrapper'>

            <div className='header-card'>
                <div className='title-card'> { shipName } </div>
            </div>

            <div className='starship-image-wrapper'>
              <img className='starship-image' src={ imagePath } onError={change2DefaultImage}/>
            </div>

            <div className='data-card'>
                <p>
                  <b>Cargo capacity</b>: {formatBigNumbers(shipCapacity)}
                </p>
                <p>
                  <b>MGLT</b>: {MGLT}
                </p>
                <p>
                  <b>Consumables</b>: {consumables}
                </p>
                <p>
                  <b>Cost in credits</b>: {formatBigNumbers(costInCredits)}
                </p>
                <p>
                  <b>Crew</b>: {crew}
                </p>
                <p>
                  <b>Hyperdriver Rating</b>: {hyperDriverRating}
                </p>
                <p>
                  <b>Manufacturer</b>: {manufacturer}
                </p>
            </div>
        </div>
    </div>
  )
}

StarshipCard.propTypes = {
  shipName: PropTypes.string.isRequired, 
  shipCapacity: PropTypes.string.isRequired,
  MGLT: PropTypes.string.isRequired, 
  consumables: PropTypes.string.isRequired, 
  costInCredits: PropTypes.string.isRequired, 
  crew: PropTypes.string.isRequired, 
  hyperDriverRating: PropTypes.string.isRequired, 
  manufacturer: PropTypes.string.isRequired
}
