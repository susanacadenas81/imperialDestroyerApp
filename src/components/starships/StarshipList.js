import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import { getSortData } from '../../helpers/getData';
import { useFetchData } from '../../hooks/useFetchData';
import { StarshipCard } from './StarshipCard';
import { Paginated } from '../ui/Paginated';

//import '../styles/components.css';
import './starships.css';

export const StarshipList = ( sortBy ) => {
  const { count, data, loading } = useFetchData('starships');
  const [paginatedData, setPaginatedData] = useState([]);
  const onChangePage = (paginatedData) => {
    setPaginatedData(paginatedData)
  }
  let sortData = [];
  useEffect(() => {
    if ( !loading ) {
      localStorage.setItem('starships', JSON.stringify(data));
      sortData = sortBy ? getSortData(data, sortBy) : data;
    }
  }, [loading, sortBy]);
    
    return (
      <>
      <div className='list-wrapper'>  
          {
              paginatedData.map((starship) => <StarshipCard 
                key={ starship.url } 
                shipName={ starship.name }
                shipCapacity={ starship.cargo_capacity }
                MGLT={ starship.MGLT }
                consumables={starship.consumables}
                costInCredits={starship.cost_in_credits}
                crew={starship.crew}
                hyperDriverRating={starship.hyperdrive_rating}
                manufacturer={starship.manufacturer}
              />)
              
          }
      </div>

      <Paginated 
        onChangePage={ onChangePage } 
        data={ data } 
        sortBy={ sortBy } 
        count={ count }
        loading={ loading }
      />

    </>
    )
}

StarshipList.propTypes = {
  sortBy: PropTypes.string.isRequired 
}
