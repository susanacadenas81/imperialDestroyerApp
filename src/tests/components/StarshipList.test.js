import { shallow } from 'enzyme'
import '@testing-library/jest-dom';
import { StarshipList } from '../../components/starships/StarshipList';
import { useFetchData } from '../../hooks/useFetchData';

jest.mock('../../hooks/useFetchData');

describe('Tests StarshipList component', () => {
    test('Should display correctly', () => {
        useFetchData.mockReturnValue({
            count: 0,
            data: [],
            loading: true
        })
        const wrapper = shallow(<StarshipList sortBy='' />);
        expect( wrapper ).toMatchSnapshot();
    })

    test('Should show items when data are loaded', () => {
        const starships = [
            {
                "name": "CR90 corvette",
                "model": "CR90 corvette",
                "manufacturer": "Corellian Engineering Corporation",
                "cost_in_credits": "3500000",
                "length": "150",
                "max_atmosphering_speed": "950",
                "crew": "30-165",
                "passengers": "600",
                "cargo_capacity": "3000000",
                "consumables": "1 year",
                "hyperdrive_rating": "2.0",
                "MGLT": "60",
                "starship_class": "corvette",
                "pilots": [],
                "films": [
                    "https://swapi.py4e.com/api/films/1/",
                    "https://swapi.py4e.com/api/films/3/",
                    "https://swapi.py4e.com/api/films/6/"
                ],
                "created": "2014-12-10T14:20:33.369000Z",
                "edited": "2014-12-20T21:23:49.867000Z",
                "url": "https://swapi.py4e.com/api/starships/2/"
            },
            {
                "name": "Star Destroyer",
                "model": "Imperial I-class Star Destroyer",
                "manufacturer": "Kuat Drive Yards",
                "cost_in_credits": "150000000",
                "length": "1,600",
                "max_atmosphering_speed": "975",
                "crew": "47,060",
                "passengers": "n/a",
                "cargo_capacity": "36000000",
                "consumables": "2 years",
                "hyperdrive_rating": "2.0",
                "MGLT": "60",
                "starship_class": "Star Destroyer",
                "pilots": [],
                "films": [
                    "https://swapi.py4e.com/api/films/1/",
                    "https://swapi.py4e.com/api/films/2/",
                    "https://swapi.py4e.com/api/films/3/"
                ],
                "created": "2014-12-10T15:08:19.848000Z",
                "edited": "2014-12-20T21:23:49.870000Z",
                "url": "https://swapi.py4e.com/api/starships/3/"
            }
        ]
        useFetchData.mockReturnValue({
            count: starships.length,
            data: starships,
            loading: false
        })
        const wrapper = shallow(<StarshipList sortBy='' />);
        expect ( wrapper ).toMatchSnapshot();
    })
})