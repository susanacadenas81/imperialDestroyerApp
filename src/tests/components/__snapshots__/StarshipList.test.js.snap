// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests StarshipList component Should display correctly 1`] = `
<Fragment>
  <div
    className="list-wrapper"
  />
  <Paginated
    count={0}
    data={Array []}
    loading={true}
    onChangePage={[Function]}
    sortBy={
      Object {
        "sortBy": "",
      }
    }
  />
</Fragment>
`;

exports[`Tests StarshipList component Should show items when data are loaded 1`] = `
<Fragment>
  <div
    className="list-wrapper"
  />
  <Paginated
    count={2}
    data={
      Array [
        Object {
          "MGLT": "60",
          "cargo_capacity": "3000000",
          "consumables": "1 year",
          "cost_in_credits": "3500000",
          "created": "2014-12-10T14:20:33.369000Z",
          "crew": "30-165",
          "edited": "2014-12-20T21:23:49.867000Z",
          "films": Array [
            "https://swapi.py4e.com/api/films/1/",
            "https://swapi.py4e.com/api/films/3/",
            "https://swapi.py4e.com/api/films/6/",
          ],
          "hyperdrive_rating": "2.0",
          "length": "150",
          "manufacturer": "Corellian Engineering Corporation",
          "max_atmosphering_speed": "950",
          "model": "CR90 corvette",
          "name": "CR90 corvette",
          "passengers": "600",
          "pilots": Array [],
          "starship_class": "corvette",
          "url": "https://swapi.py4e.com/api/starships/2/",
        },
        Object {
          "MGLT": "60",
          "cargo_capacity": "36000000",
          "consumables": "2 years",
          "cost_in_credits": "150000000",
          "created": "2014-12-10T15:08:19.848000Z",
          "crew": "47,060",
          "edited": "2014-12-20T21:23:49.870000Z",
          "films": Array [
            "https://swapi.py4e.com/api/films/1/",
            "https://swapi.py4e.com/api/films/2/",
            "https://swapi.py4e.com/api/films/3/",
          ],
          "hyperdrive_rating": "2.0",
          "length": "1,600",
          "manufacturer": "Kuat Drive Yards",
          "max_atmosphering_speed": "975",
          "model": "Imperial I-class Star Destroyer",
          "name": "Star Destroyer",
          "passengers": "n/a",
          "pilots": Array [],
          "starship_class": "Star Destroyer",
          "url": "https://swapi.py4e.com/api/starships/3/",
        },
      ]
    }
    loading={false}
    onChangePage={[Function]}
    sortBy={
      Object {
        "sortBy": "",
      }
    }
  />
</Fragment>
`;
